*: add slow log control variables (`tidb_slow_log_rules`, `tidb_slow_log_max_per_sec`) and hint `WRITE_SLOW_LOG` (#22176)
pingcap/docs: identify-slow-queries.md

- `Storage_from_kv`: introduced in v9.0.0, indicates whether this statement read data from TiKV.
- `Storage_from_mpp`: introduced in v9.0.0, indicates whether this statement read data from TiFlash.

## Related hint

> You can force a statement to be recorded in the slow query log by using the `WRITE_SLOW_LOG` hint.

Author: ```suggestion
Starting from v9.0.0, you can use the `WRITE_SLOW_LOG` hint to force TiDB to output the execution information of a specific SQL statement to the slow query log, regardless of whether its execution time exceeds the threshold. This is useful for capturing detailed metadata during SQL execution, such as execution plans and resource consumption. For example, when you troubleshoot intermittent performance issues, some SQL statements might slow down only under specific conditions, such as sudden data spikes or index failures, while performing quickly at other times, making them difficult to capture using the default threshold. By pre-emptively adding this hint to business SQL queries, you ensure that the system captures a complete record of detailed information the moment the issue recurs.
```

- `Storage_from_kv`: introduced in v9.0.0, indicates whether this statement read data from TiKV.
- `Storage_from_mpp`: introduced in v9.0.0, indicates whether this statement read data from TiFlash.

## Related hint

> Starting from v9.0.0, you can use the `WRITE_SLOW_LOG` hint to force TiDB to output the execution information of a specific SQL statement to the slow query log, regardless of whether its execution time exceeds the threshold. This is useful for capturing detailed metadata during SQL execution, such as execution plans and resource consumption. For example, when you troubleshoot intermittent performance issues, some SQL statements might slow down only under specific conditions, such as sudden data spikes or index failures, while performing quickly at other times, making them difficult to capture using the default threshold. By pre-emptively adding this hint to business SQL queries, you ensure that the system captures a complete record of detailed information the moment the issue recurs.

Reviewer: ```suggestion
Starting from v9.0.0, you can use the `WRITE_SLOW_LOG` hint to force TiDB to output the execution information of a specific SQL statement to the slow query log, regardless of whether its execution time exceeds the threshold. This is useful for capturing detailed metadata during SQL execution, such as execution plans and resource consumption.

For example, when you troubleshoot intermittent performance jitters, some SQL statements might slow down only under specific conditions, such as sudden data spikes or index failures, while performing quickly under normal circumstances, making them difficult to capture using the default threshold. In such scenarios, you can proactively add the `WRITE_SLOW_LOG` hint to your application's SQL in advance. This ensures that the system records the complete execution details the moment the issue recurs.
```

- This hint bypasses any thresholds or trigger rules, which means the statement is always written to the slow query log.

> - Currently, you can only force a statement to be written to the slow query log. Disabling this behavior with `WRITE_SLOW_LOG(FALSE)` is not supported.

Reviewer: ```suggestion
- The `WRITE_SLOW_LOG` hint currently only supports forcing a statement to be recorded in the slow query log. Disabling log recording using `WRITE_SLOW_LOG(FALSE)` is not supported.
```

            * TiKV Coprocessor Task-related fields:
                * `Process_time`, `Total_keys`, `Process_keys`, `Num_cop_tasks`
            * Memory usage related fields:
                * `Mem_max`
            * Disk usage related fields:
                * `Disk_max`
            * Resource control related fields:
                * `Resource_group`
            * Network transmission related fields:
                * `KV_total`, `PD_total`
                * `Unpacked_bytes_sent_tikv_total`, `Unpacked_bytes_received_tikv_total`
                * `Unpacked_bytes_sent_tikv_cross_zone`, `Unpacked_bytes_received_tikv_cross_zone`
                * `Unpacked_bytes_sent_tiflash_total`, `Unpacked_bytes_received_tiflash_total`
                * `Unpacked_bytes_sent_tiflash_cross_zone`, `Unpacked_bytes_received_tiflash_cross_zone`
    * If `tidb_slow_log_rules` is not set:
        * Slow query logging still relies on `tidb_slow_log_threshold`. The `query_time` threshold is taken from that variable for backward compatibility.
    * If `tidb_slow_log_rules` is set:
        * Configured rules take precedence and `tidb_slow_log_threshold` is ignored.

>         * If you want `query_time` to be part of the trigger, include it explicitly in the rules.

Reviewer: ```suggestion
        * If you want to use SQL execution time as a condition for slow log output, configure the rule using `query_time` with a defined threshold.
```

        * Rule matching logic (multiple rules use OR):

>             * SESSION-scoped rules are matched first; if a SESSION rule matches, the slow query is logged.

Reviewer: ```suggestion
            * SESSION scope rules are matched first; if a SESSION rule matches, the slow query is logged.
```

            * GLOBAL-scoped rules are considered only when no SESSION rule matches:
                * If a GLOBAL rule specifies `ConnID` and it matches the current session `ConnID`, that rule takes effect.

>                 * If a GLOBAL rule does not specify `ConnID` (a global rule), that rule takes effect.

Reviewer: ```suggestion
            * GLOBAL scoped rules are evaluated only when no SESSION scope rules match:
                * If a GLOBAL rule specifies a `ConnID` that matches the current session `ConnID`, that rule is applied.
                * If a GLOBAL rule does not specify a `ConnID` (acting as a general global rule), that rule is applied.
```

>       * The behavior of `SHOW VARIABLES`, `SELECT @@GLOBAL.tidb_slow_log_rules`, and `SELECT @@SESSION.tidb_slow_log_rules` for this variable is the same as for other system variables.

Reviewer: ```suggestion
      * The behavior of displaying this variable via `SHOW VARIABLES`, `SELECT @@GLOBAL.tidb_slow_log_rules`, or `SELECT @@SESSION.tidb_slow_log_rules` is consistent with other system variables.
```

    * Examples:
        * Standard format (SESSION scope):

          ```sql
          SET SESSION tidb_slow_log_rules = 'Query_time: 500, Is_internal: false';
          ```

        * Incorrect format (SESSION scope does not support `ConnID`):

          ```sql
          SET SESSION tidb_slow_log_rules = 'ConnID: 12, Query_time: 500, Is_internal: false';
          ```

        * Global rules (applies to all connections):

          ```sql
          SET GLOBAL tidb_slow_log_rules = 'Query_time: 500, Is_internal: false';
          ```

>         * Global rules for specific connections (applies to connections with `ConnID: 11` and `ConnID: 12`, respectively):

Reviewer: ```suggestion
        * Global rules for specific connections (applying to connections with `ConnID: 11` and `ConnID: 12` respectively):
```

    > - In a well-provisioned test environment (1 TiDB node: 16C/48G, and 3 TiKV nodes: 16C/48G), multiple rounds of sysbench tests show that when multi-dimensional rules generate millions of slow query log entries within half an hour, the performance impact is small. However, when log volume reaches tens of millions, TPS and latency degrade noticeably. On high-workload systems or when CPU/memory are near capacity, configure `tidb_slow_log_rules` conservatively to avoid log floods. It is recommended to use `tidb_slow_log_max_per_sec` to throttle slow query log printing and reduce impact on production workloads.

* [`tidb_slow_log_max_per_sec`](/system-variables.md#tidb_slow_log_max_per_sec-new-in-v900): sets the maximum number of slow query logs printed per second. This variable is introduced in v9.0.0. The default value is `0`.
    * A value of `0` means there is no limit on the number of slow query logs printed per second.
    * A value greater than `0` caps the number of slow query logs printed per second. Excess logs are discarded and not written to the slow query log file.

>     * It is recommended to set this variable when `tidb_slow_log_rules` is enabled to avoid excessive log printing.

Reviewer: ```suggestion
    * A value of `0` means there is no limit on the number of slow query log entries written per second.
    * A value greater than `0` means TiDB writes at most the specified number of slow query log entries per second. Any excess log entries are discarded and not written to the slow query log file.
    * It is recommended to set this variable after enabling `tidb_slow_log_rules` to avoid frequent log writes triggered by rules.
```

* [`tidb_slow_log_rules`](/system-variables.md#tidb_slow_log_rules-new-in-v900): defines trigger rules for the slow query log. It supports combining multi-dimensional metrics to provide more flexible and fine-grained logging. Introduced in v9.0.0, it gradually replaces the single-threshold approach (`tidb_slow_log_threshold`). This variable supports using the following fields as filter conditions when slow query logs are output. For more information, see [Fields description](#fields-description).
    * Supported filter fields:
        * Basic slow query information:
            * `Query_time`, `Parse_time`, `Compile_time`, `Optimize_time`, `Wait_TS`, `Rewrite_time`
            * `Digest`, `Plan_digest`, `Is_internal`, `Succ`
            * `Exec_retry_count`, `Backoff_time`, `Write_sql_response_total`
        * Transaction-related fields:
            * `Prewrite_time`, `Commit_time`, `Write_keys`, `Write_size`, `Prewrite_region`
        * User-related fields for SQL execution:
            * `Conn_ID`, `DB`, `Session_alias`

>         * TiKV Coprocessor Task-related fields:

Reviewer: ```suggestion
        * TiKV Coprocessor task-related fields:
```
